### PR TITLE
#2282 Generic filters (cold chain) -- Part 1

### PR DESCRIPTION
**Pull request by CarlosNZ**
First part of #2282.

Have added a "Progress" list to the [original issue](#2282) to keep track of where this up to.

# 👩🏻‍💻 What does this PR do? 
 

Have created the basic mechanism, and the first two filter types (Text and Enum). But thought it'd be good to get some feedback/agreement on some of the UI/UX aspects before doing the rest, especially since these filters will probably be used in quite a few places eventually. (It took a long time to figure out all the UI quirks to make these two filters visually consistent, so would be good to lock it down before going further)

How it works is there is a "FilterRoot" component, to which you pass all the definitions for the filters you want to make available. See code for data shape.

Currently, I've implemented them for the "Patients" list (/dispensary/patients?sort=code) and replaced the three "SearchBar" filters there with this single "FilterRoot" component.

The "Gender" field is not currently filterable, but I've left a filter definition for a "Gender" enum commented out, so you can test it and see how it looks, even though it won't do anything (it'll update the url, but won't affect search results)

Initial filter options:
Screenshot 2023-10-11 at 1 33 20 PM

After applying a couple:
Screenshot 2023-10-11 at 1 33 38 PM

## 💌 Any notes for the reviewer?

Note that the filters are re-populated when loading the page when query terms are in the URL (refresh the page -- filters should be preserved)

### Some considerations:

- Currently, the "X" icon will remove the filter completely. I originally had it so that it would just clear the input, but then  the only way to remove them is to use the "Clear all filters" option. We could add a different icon outside the input for removing, but then I think the UI will get a little cluttered. There's pros and cons to both approaches, so see what you think (or suggest an alternative)
- The "loading" spinner you see in the Text input is a bit misleading (IMHO). It actually just shows the debounce waiting, not the search request. This is the same as in the current "SearchBar" component -- just wondering what people think. Do we need it all?
- I've put these into the "common" package, as I presumed these would be generic enough to use throughout. Is this correct?
- See inline comments for more detail

### CODE REVIEW

File: client/packages/common/src/ui/components/inputs/Filters/EnumFilter.tsx [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+          zIndex: 100,
+          top: '4px',
+          left: '8px',
+          color: 'gray.main',
+        },
+        '& .MuiInputLabel-root.Mui-focused': {
+          color: 'secondary.main',
+        },
+      }}
+      label={name}
```
Comment (CarlosNZ) [inline client/packages/common/src/ui/components/inputs/Filters/EnumFilter.tsx:56]: This is repeated (and sourced from elsewhere). Will pull into a common file once all filters are done.

File: client/packages/common/src/ui/components/inputs/Filters/FilterRoot.tsx [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  const removeFilter = (filterDefinition: FilterDefinition) => {
+    const newActiveFilters = activeFilters.filter(
+      fil => fil.urlParameter !== filterDefinition.urlParameter
+    );
+    updateQuery({ [filterDefinition.urlParameter]: '' });
+    setActiveFilters(newActiveFilters);
+  };
+
+  return (
+    <Box display="flex" gap={2} sx={{ alignItems: 'flex-end', minHeight: 50 }}>
```
Comment (CarlosNZ) [inline client/packages/common/src/ui/components/inputs/Filters/FilterRoot.tsx:64]: `minHeight` prevent a UI "jump" when instantiating the first filter, as those components are a bit taller than this root component.

File: client/packages/common/src/ui/components/inputs/Filters/FilterRoot.tsx [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    <InputAdornment position="end">
+      <IconButton
+        sx={{ color: 'gray.main' }}
+        label={hasValue ? t('label.clear-filter') : ''}
+        onClick={hasValue ? onClear : () => {}}
+        icon={hasValue ? <CloseIcon /> : <SearchIcon fontSize="small" />}
+      />
+    </InputAdornment>
+  );
+};
```
Comment (CarlosNZ) [inline client/packages/common/src/ui/components/inputs/Filters/FilterRoot.tsx:146]: Also copied from elsewhere, can probably consolidate at some point.

File: client/packages/system/src/Patient/ListView/Toolbar.tsx [hunk 4]
```diff
@@ -33,26 +20,36 @@ export const Toolbar: FC<{
...
+            },
+            // {
+            //   type: 'enum',
+            //   name: 'Gender',
+            //   urlParameter: 'gender',
+            //   options: [
+            //     { label: 'Male', value: 'MALE' },
+            //     { label: 'Female', value: 'FEMALE' },
+            //   ],
+            // },
```
Comment (CarlosNZ) [inline client/packages/system/src/Patient/ListView/Toolbar.tsx:51]: Uncomment this to test out the Enum filter

**Comment by mark-prins** [comment]
That's awesome! Great to see the filters take shape - I've wanted them in place for a long time.
Some replies to your questions in the description and a nit picky observation:

**Clear filter**

Two things - the wording and the expectation of what happens. For me, `clear` means that the input remains but is 'cleared'. The `X` icon is used in the search bar currently to clear the input, not remove it; so my expectation is that the input is cleared.

Screenshot 2023-10-12 at 9 03 47 AM

What I would prefer:

- the 'X' clears the input and the filter is not removed
- the main filter Select has the final option as `Remove all filters` and not `Clear all filters`

I don't see the need to have a 'remove filter' option on the specific filters - clearing them will remove the effect of the filter anyway, so there's no practical need to have the input removed. and that is possible by going back to the component which controls which filters are shown.

**Naming**

This is very nit picky! I'm slightly confused by the name `FilterRoot` because I'm not sure what the 'Root' is intending. I thought when I first read the name that it was a `BaseFilter` i.e. something to be extended, but that doesn't appear to be the case. It could just be `Filter` and give me the same amount of meaning

**Package**

Agree, the common package is best for it. The filter will be shared by a number of the client packages

**Loader**

I rather like having the feedback as I type, though yeah it is misleading! I've now taken the loader off and it's not clear that something is happening when I type. It also makes the reloading of the list view look clumsier - perhaps I am just noticing it more now! Ideally - the list view has a loader / some sort of transition / animation; though the search is so quick it probably won't have time for a loader. 

So. conflicted. Ideally the loader will show as an end adornment when the search is happening but that requires passing the `isLoading` to the filter component and doing a whole lot of work which I don't think we should do. My vote is to remove the loading indicator from the inputs.

File: client/packages/common/src/ui/components/inputs/Filters/FilterRoot.tsx [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    const newActiveFilters = activeFilters.filter(
+      fil => fil.urlParameter !== filterDefinition.urlParameter
+    );
+    updateQuery({ [filterDefinition.urlParameter]: '' });
+    setActiveFilters(newActiveFilters);
+  };
+
+  return (
+    <Box display="flex" gap={2} sx={{ alignItems: 'flex-end', minHeight: 50 }}>
+      <DropdownMenu label="Filters">
```
Comment (mark-prins) [inline client/packages/common/src/ui/components/inputs/Filters/FilterRoot.tsx:65]: needs to have a translation 😉

File: client/packages/common/src/ui/components/inputs/Filters/FilterRoot.tsx [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        {filterOptions.map(option => (
+          <DropdownMenuItem
+            key={option.value}
+            onClick={() => handleSelect(option.value)}
+            sx={{ fontSize: 14 }}
```
Comment (mark-prins) [inline client/packages/common/src/ui/components/inputs/Filters/FilterRoot.tsx:70]: perhaps could be a CONST as well, since used in two places?

**Review by mark-prins — CHANGES_REQUESTED** [review]
🎉  that's very nice! like how good it is to use and how easy to add to an existing area.
Just a few comments, nothing major to change I think

File: client/packages/system/src/Patient/ListView/Toolbar.tsx [hunk 7]
```diff
@@ -33,26 +20,36 @@ export const Toolbar: FC<{
...
-          }}
-        />
-        <SearchBar
-          placeholder={t('placeholder.search-by-identifier')}
-          value={identifierFilterString}
-          onChange={newValue => {
-            filter.onChangeStringFilterRule('identifier', 'like', newValue);
-          }}
+        <FilterRoot
+          filters={[
```
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply client/packages/system/src/Patient/ListView/Toolbar.tsx:24]: They appear in the order they're defined!

(You pass in an array of filter definitions, so you can put them in whatever order you like. Personally, I would have things grouped together somewhat logically -- i.e First and Last name should be next to each other. And I'd put more rarely used ones at the bottom)

File: client/packages/common/src/intl/locales/en/common.json [hunk 8]
```diff
@@ -122,6 +122,7 @@
   "label.cant-delete-disabled-requisition": "Can only delete lines when requisition is in Draft",
   "label.chargeCode": "Charge Code",
   "label.clear-filter": "Clear filter",
+  "label.clear-all-filters": "Clear all filters",
```
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply client/packages/common/src/intl/locales/en/common.json]: 👍 4f29e8cf8

File: client/packages/common/src/intl/locales/en/dispensary.json [hunk 9]
```diff
@@ -28,6 +28,9 @@
   "label.patient": "Patient",
   "label.patients": "Patients",
   "label.search-results": "Search results",
+  "label.search-first-name": "First name",
+  "label.search-last-name": "Last name",
+  "label.search-identifier": "Identifier",
```
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply client/packages/common/src/intl/locales/en/dispensary.json]: Ah yes, good point: 7ac152f4b

File: client/packages/common/src/ui/components/inputs/Filters/FilterRoot.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply client/packages/common/src/ui/components/inputs/Filters/FilterRoot.tsx]: Indeed! 
Screenshot 2023-10-12 at 2 03 53 PM

9aed0fffb

**Comment by CarlosNZ** [comment]
> **Loader**
> 
> I rather like having the feedback as I type, though yeah it is misleading! I've now taken the loader off and it's not clear that something is happening when I type. It also makes the reloading of the list view look clumsier - perhaps I am just noticing it more now! Ideally - the list view has a loader / some sort of transition / animation; though the search is so quick it probably won't have time for a loader.
> 
> So. conflicted. Ideally the loader will show as an end adornment when the search is happening but that requires passing the `isLoading` to the filter component and doing a whole lot of work which I don't think we should do. My vote is to remove the loading indicator from the inputs.

Yeah, how about removing the spinner from the input but reducing the debounce time to something closer to 200ms (It's currently 500, which seems a bit excessive)?

We already have the main spinner in the list, so it seems like we don't really need one for each search input.

**Comment by CarlosNZ** [comment]
> **Naming**
> 
> This is very nit picky! I'm slightly confused by the name `FilterRoot` because I'm not sure what the 'Root' is intending. I thought when I first read the name that it was a `BaseFilter` i.e. something to be extended, but that doesn't appear to be the case. It could just be `Filter` and give me the same amount of meaning

I've gone with `FilterMenu`, which captures the UI of what it is :) 4cfe3e350

File: client/packages/common/src/ui/components/inputs/Filters/FilterRoot.tsx (hunk 6 above)
Comment (CarlosNZ) [reply client/packages/common/src/ui/components/inputs/Filters/FilterRoot.tsx]: Okay, I've abstracted this into a separate, small `FilterMenuItem` component

1a1510a59

File: client/packages/common/src/ui/components/inputs/Filters/FilterRoot.tsx (hunk 5 above)
Comment (CarlosNZ) [reply client/packages/common/src/ui/components/inputs/Filters/FilterRoot.tsx]: 😳 84abcdb51

**Comment by CarlosNZ** [comment]
> What I would prefer:
> 
>     * the 'X' clears the input and the filter is not removed
> 
>     * the main filter Select has the final option as `Remove all filters` and not `Clear all filters`
> 
> 
> I don't see the need to have a 'remove filter' option on the specific filters - clearing them will remove the effect of the filter anyway, so there's no practical need to have the input removed. and that is possible by going back to the component which controls which filters are shown.

I'll do that to start with. 

I do kind of miss being able to remove filters without removing the whole lot -- kind of "cleaning up". (Refreshing the page is a clunky way to achieve this 😉 ). One possibility is to make the "Remove all filters" text conditional -- if there are "unused" filters, then this would say "Remove unused filters", and once there are no more unused filters it would change to "Remove all filters". Or just show both options?

What do you think? Too complicated?

**Comment by mark-prins** [comment]
> I do kind of miss being able to remove filters without removing the whole lot -- kind of "cleaning up". 

ah, right. good point, it does mean clearing them all and then re-adding which isn't ideal. I still prefer the 'individual clear' and 'remove all' options. If we do end up with individual remove, can you rename the tooltip tho, so it says 'remove' rather than 'clear'?

> What do you think? Too complicated?

yeah, I think it probably is. Too many options for users to get their heads around!

**Review by mark-prins — APPROVED** [review]
👍  thanks for the changes! I agree about removing the loader icon and changing down to 300ms, that simplifies the code a little too.

Just the adornments on the enum filter.. that's not a serious concern tho

File: client/packages/system/src/Patient/ListView/Toolbar.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply client/packages/system/src/Patient/ListView/Toolbar.tsx:24]: One thing I'd like to do (I'll probably do something like it with the "Date/Time" filters), is to be able to define a filter "group". So the Filter menu have a single entry for "Name", and when you select it both the "First Name" and "Last Name" filters get instantiated at the same time :)

**Comment by CarlosNZ** [comment]
> 👍 thanks for the changes! I agree about removing the loader icon and changing down to 300ms, that simplifies the code a little too.
> 
> Just the adornments on the enum filter.. that's not a serious concern tho

Okay, have addressed the adornments and changed the behaviour so that the filters get *cleared* rather than *removed* as discussed above: df0d98351

Had to tweak a few things to prevent some React errors (didn't like Fragments as a child of "Menu", so the "Clear all" option needed a little refactor)

**Pull request merged**